### PR TITLE
Improve performance of reduction for small number of elements to reduce for types where tree-reduction is needed

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
1. Renamed misspelled variable
2. If `reduction_nelems` is small, use `SequentialReductionKernel` for tree-reductions as it is done for atomic reduction
3. Tweak scaling down logic for moderately-sized number of elements to reduce. Closes gh-1461

   We should also use `max_wg` if the `iter_nelems` is very small (one), since choosing `max_wg` for large` iter_nelems` may lead to under-utilization of GPU.

---

- [x] Have you provided a meaningful PR description?
- [ ] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [x] Have you checked performance impact of proposed changes?
- [x] If this PR is a work in progress, are you opening the PR as a draft?

### CODE REVIEW

**Comment by antonwolfy** [comment]
The PR resolves the performance issue gh-1461, L2 norm benchmark is back to expected values. Thank you!

**Comment by ndgrigorian** [comment]
@oleksandr-pavlyk 
I realized while looking through ``reductions.hpp`` that floating-point types are not included in ``TypePairSupportDataForCompReductionAtomic``, so even though we permit atomics for ``max`` and ``min`` on floating-point arrays, there aren't any functions to call in the table.

``float`` and ``double`` should be added to fix this.

File: dpctl/tensor/_elementwise_common.py [hunk 1]
```diff
@@ -363,6 +411,56 @@ def _get_shape(o):
...
+    Args:
+        name (str):
+            Name of the unary function
+        result_type_resovle_fn (callable):
+            Function that takes dtypes of the input and
+            returns the dtype of the result if the
+            implementation functions supports it, or
+            returns `None` otherwise.
+        binary_dp_impl_fn (callable):
+            Data-parallel umplementation function with signature
```
Comment (ndgrigorian) [inline dpctl/tensor/_elementwise_common.py:424]: Nit
```suggestion
            Data-parallel implementation function with signature
```

File: dpctl/tensor/_elementwise_common.py [hunk 2]
```diff
@@ -363,6 +411,56 @@ def _get_shape(o):
...
+            The first event corresponds to data-management host tasks,
+            including lifetime management of argument Python objects to ensure
+            that their associated USM allocation is not freed before offloaded
+            computational tasks complete execution, while the second event
+            corresponds to computational tasks associated with function
+            evaluation.
+        docs (str):
+            Documentation string for the unary function.
+        binary_inplace_fn (callable, optional):
+            Data-parallel omplementation function with signature
```
Comment (ndgrigorian) [inline dpctl/tensor/_elementwise_common.py:440]:
```suggestion
            Data-parallel implementation function with signature
```

**Comment by ndgrigorian** [comment]
https://github.com/IntelPython/dpctl/blob/af28d981d513729a31349fe19cb1da10515ab849/dpctl/tensor/libtensor/source/tensor_ctors.cpp#L40
https://github.com/IntelPython/dpctl/blob/af28d981d513729a31349fe19cb1da10515ab849/dpctl/tensor/libtensor/source/tensor_ctors.cpp#L51

These includes in ``tensor_ctors.cpp`` should be unnecessary with these changes.

**Comment by oleksandr-pavlyk** [comment]
> dpctl/dpctl/tensor/libtensor/source/tensor_ctors.cpp

Thank you @ndgrigorian, good catch!

**Comment by oleksandr-pavlyk** [comment]
Closes gh-1461

**Review by ndgrigorian — APPROVED** [review]
Thank you @oleksandr-pavlyk, LGTM

**Pull request merged**